### PR TITLE
Add simple percentage based TraceProfilingScopeManager

### PR DESCRIPTION
**Pull request by tylerbenson**
And associated config.  If not configured, the heuristical model is used.

Also add some accessor methods to the AgentSpan interface to gain access to the trace/span ids.

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/scopemanager/TraceProfilingScopeManager.java [hunk 1]
```diff
@@ -21,62 +23,106 @@ protected Boolean initialValue() {
...
+      return new TraceProfilingScope(delegate.handleSpan(span));
     }
 
-    if (lastProfileTimestamp + MAX_NANOSECONDS_BETWEEN_ACTIVATIONS < System.nanoTime()) {
-      // It's been a while since we last profiled, so lets take one now.
-      // Due to the multi-threaded nature here, we will likely have multiple threads enter here
-      // but they will still be subject to the rate limiter following.
-      return true;
+    private boolean shouldSample(final AgentSpan span) {
+      return span.getTraceId().compareTo(cutoff) <= 0;
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/scopemanager/TraceProfilingScopeManager.java:66]: This is a Bernoulli trial - you evaluate this once condition once for every span and should sample roughly `percent` of the time. What you _could_ do to eliminate any overhead except when you profile is generate the number of unsuccessful trials - the number of spans you won't sample - there should be between two sampled spans. 

The number of unsuccessful trials between two successes is geometrically distributed and you can approximate it with `Math.log(ThreadLocalRandom.current().nextDouble())/Math.log(1 - percent) + 1`.

Concretely, introduce variables `long nextSampledSpan = Math.log(ThreadLocalRandom.current().nextDouble())/Math.log(1 - percent) + 1` and `long spanCount = 0`. On each span, increment `spanCount`. If `spanCount == nextSampledSpan`, sample and update (`nextSampledSpan += Math.log(ThreadLocalRandom.current().nextDouble())/Math.log(1 - percent) + 1`). If `nextSampledSpan != spanCount`, you do literally nothing except bump a counter; the overhead is tiny unless you're actually going to do some useful work (activate the profiler).

Comment (jbachorik) [reply dd-trace-core/src/main/java/datadog/trace/core/scopemanager/TraceProfilingScopeManager.java:66]: 👍

Comment (tylerbenson) [reply dd-trace-core/src/main/java/datadog/trace/core/scopemanager/TraceProfilingScopeManager.java:67]: I think this is a good idea and something we should revisit if we decide to keep this code.  For now I can't implement it with minimal changes and maintain the consistent trace evaluation feature by using the trace id. I don't expect this class to be widely used in the public, so I'd rather keep the impact small.

**Comment by tylerbenson** [comment]
@randomanderson `handleSpan` is coming from `ScopeInterceptor`. But you're right, I could override it in `TraceProfilingScopeManager` and simplify the logic some.  I'll go ahead and do that.  Thanks for the idea.

**Pull request merged**